File: foundational_design_patterns/12_computer_use/src/computer_use_advanced.py
"""Computer Use advanced: optional Playwright execution + safety policy."""

from __future__ import annotations

import os
import sys

from pathlib import Path

ROOT_DIR = next(parent for parent in Path(__file__).resolve().parents if (parent / "ssl_fix.py").exists())
if str(ROOT_DIR) not in sys.path:
    sys.path.insert(0, str(ROOT_DIR))

from repo_support import configure_example, get_default_model

configure_example(__file__)

from langchain_openai import ChatOpenAI


def run_playwright_task() -> str:
    try:
        from playwright.sync_api import sync_playwright
    except Exception:
        return "Playwright not installed; skipping real browser automation."

    with sync_playwright() as p:
        browser = p.chromium.launch(headless=True)
        page = browser.new_page()
        page.goto("https://en.wikipedia.org/wiki/Artificial_intelligence", wait_until="domcontentloaded")
        title = page.title()
        heading = page.locator("#firstHeading").inner_text()
        browser.close()
    return f"Visited page title='{title}', heading='{heading}'"


def main() -> None:
    print("\n" + "=" * 80)
    print("COMPUTER USE — ADVANCED")
    print("=" * 80)

    safety_policy = {
        "allowlist_domains": ["wikipedia.org"],
        "blocked_actions": ["checkout", "payment", "account deletion"],
        "human_approval_required_for": ["form submit", "external navigation", "file download"],
    }
    print("\nSafety policy:")
    print(safety_policy)

    outcome = run_playwright_task()
    print("\nAutomation outcome:")
    print(outcome)

    if os.getenv("OPENAI_API_KEY"):
        try:
            llm = ChatOpenAI(model=get_default_model(), temperature=0)
            summary = llm.invoke(f"Summarize this browser automation run in 3 bullets:\n{outcome}").content
            print("\nLLM summary:")
            print(summary)
        except Exception as exc:
            print("\nLLM summary unavailable:")
            print(exc)


if __name__ == "__main__":
    main()

File: foundational_design_patterns/12_computer_use/typescript/src/computer_use_advanced.ts
/**
 * Computer Use advanced — TypeScript port of src/computer_use_advanced.py.
 */

import { ChatOpenAI } from "@langchain/openai";

export async function runPlaywrightTask(): Promise<string> {
  try {
    const importer = new Function(
      "specifier",
      "return import(specifier);",
    ) as (specifier: string) => Promise<{
      chromium: {
        launch(options: { headless: boolean }): Promise<{
          newPage(): Promise<{
            goto(url: string, options: { waitUntil: string }): Promise<void>;
            title(): Promise<string>;
            locator(selector: string): { innerText(): Promise<string> };
          }>;
          close(): Promise<void>;
        }>;
      };
    }>;
    const playwright = await importer("playwright");
    const browser = await playwright.chromium.launch({ headless: true });
    const page = await browser.newPage();
    await page.goto("https://en.wikipedia.org/wiki/Artificial_intelligence", {
      waitUntil: "domcontentloaded",
    });
    const title = await page.title();
    const heading = await page.locator("#firstHeading").innerText();
    await browser.close();
    return `Visited page title='${title}', heading='${heading}'`;
  } catch {
    return "Playwright not installed; skipping real browser automation.";
  }
}

async function main(): Promise<void> {
  console.log("\n" + "=".repeat(80));
  console.log("COMPUTER USE — ADVANCED");
  console.log("=".repeat(80));

  const safetyPolicy = {
    allowlist_domains: ["wikipedia.org"],
    blocked_actions: ["checkout", "payment", "account deletion"],
    human_approval_required_for: [
      "form submit",
      "external navigation",
      "file download",
    ],
  };

  console.log("\nSafety policy:");
  console.log(safetyPolicy);

  const outcome = await runPlaywrightTask();
  console.log("\nAutomation outcome:");
  console.log(outcome);

  if (process.env.OPENAI_API_KEY) {
    try {
      const llm = new ChatOpenAI({
        model: process.env.OPENAI_MODEL ?? "gpt-4o-mini",
        temperature: 0,
      });
      const response = await llm.invoke(
        `Summarize this browser automation run in 3 bullets:\n${outcome}`,
      );
      console.log("\nLLM summary:");
      console.log(response.content);
    } catch (exc: unknown) {
      console.log("\nLLM summary unavailable:");
      console.log(exc);
    }
  }
}

if (import.meta.main) {
  await main();
}
